File: lib/turbofan/runtime/metrics.rb
# frozen_string_literal: true

require "aws-sdk-cloudwatch"

module Turbofan
  module Runtime
    class Metrics
      def initialize(pipeline_name:, stage:, step_name:, cloudwatch_client: nil, size: nil)
        @cloudwatch_client = cloudwatch_client
        @pipeline_name = pipeline_name
        @stage = stage
        @step_name = step_name
        @size = size
        @pending = []
        # fan_out workers may call emit() concurrently. Array#<< is not
        # thread-safe under MRI — two racing appends can drop an entry
        # or leave the array in an inconsistent state. flush() reads and
        # drains @pending, so it also needs mutex protection or a
        # concurrent emit() can land in a batch that's already been
        # serialized to CloudWatch payload form but not yet shifted off.
        @mutex = Mutex.new
      end

      def emit(name, value, unit: nil)
        raise ArgumentError, "metric value must be Numeric, got #{value.class}" unless value.is_a?(Numeric)
        entry = {name: name, value: value}
        entry[:unit] = unit if unit
        @mutex.synchronize { @pending << entry }
      end

      # CloudWatch PutMetricData supports up to 1000 metrics per call (1 MB
      # payload cap). Previously we used 20, causing 50× more API calls than
      # needed and increased throttle exposure at high fan-out scale.
      BATCH_SIZE = 100

      def flush
        return if @pending.empty?

        # Clear each batch only after its PUT succeeds. If a later batch fails
        # after retry exhaustion, the unsent remainder stays in @pending so a
        # subsequent flush call (if the container lives long enough) can retry.
        until @pending.empty?
          batch = @mutex.synchronize { @pending.first(BATCH_SIZE) }
          metric_data = batch.map { |entry| build_metric_datum(entry) }
          Turbofan::Retryable.call do
            cloudwatch_client.put_metric_data(
              namespace: "Turbofan/#{@pipeline_name}",
              metric_data: metric_data
            )
          end
          @mutex.synchronize { @pending.shift(batch.size) }
        end
      rescue Aws::Errors::ServiceError => e
        remaining = @mutex.synchronize { @pending.size }
        warn("[Turbofan] WARNING: Failed to flush #{remaining} remaining metrics: #{e.message}")
        # Intentionally do NOT clear @pending — if this Metrics instance is
        # flushed again, we get another chance. Container teardown is the
        # only path that truly drops them.
      end

      private

      # Disable SDK's built-in retry so Turbofan::Retryable owns all retry
      # decisions when flush wraps put_metric_data. See lib/turbofan/retryable.rb.
      # `max_attempts: 1` works across standard/adaptive/legacy modes; the
      # legacy-only `retry_limit: 0` would be ignored in modern modes.
      def cloudwatch_client
        return @cloudwatch_client if defined?(@cloudwatch_client) && @cloudwatch_client
        @mutex.synchronize do
          @cloudwatch_client ||= Aws::CloudWatch::Client.new(retry_mode: "standard", max_attempts: 1)
        end
      end

      def build_metric_datum(entry)
        datum = {
          metric_name: entry[:name],
          value: entry[:value],
          dimensions: dimensions
        }
        datum[:unit] = entry[:unit] if entry[:unit]
        datum
      end

      def dimensions
        dims = [
          {name: "Pipeline", value: @pipeline_name},
          {name: "Stage", value: @stage},
          {name: "Step", value: @step_name}
        ]
        dims << {name: "Size", value: @size.to_s} if @size
        dims
      end
    end
  end
end

File: python/turbofan_runtime/metrics.py
"""CloudWatch metrics emission — batched, thread-safe, retried.

Mirrors lib/turbofan/runtime/metrics.rb. Namespace `Turbofan/{pipeline}`,
dimensions `Pipeline`/`Stage`/`Step` (+ optional `Size`). Per-emit
appends to a mutex-protected pending list; flush() drains in batches
of 100 (the CloudWatch PutMetricData payload limit is 1000 / 1 MB but
100 keeps individual payloads small and limits per-call blast radius
on partial failures).

flush() is wrapped in `Retryable.call(..., max_retry_seconds=None)`
because it's typically called from the wrapper's `finally` block
during a SIGTERM unwind — if the global retry budget aborted the
flush, we'd lose the very telemetry that records the failure. Same
rationale as `OutputSerializer` and `Payload.serialize`.
"""

import sys
import threading

import boto3
from botocore.config import Config

from .retryable import Retryable


class Metrics:
    BATCH_SIZE = 100  # mirrors Ruby Metrics::BATCH_SIZE

    def __init__(
        self,
        *,
        pipeline_name,
        stage,
        step_name,
        size=None,
        cloudwatch_client=None,
    ):
        self._pipeline_name = pipeline_name
        self._stage = stage
        self._step_name = step_name
        self._size = size
        self._cw = cloudwatch_client
        self._pending = []
        self._lock = threading.Lock()

    def emit(self, name, value, unit=None):
        if not isinstance(value, (int, float)) or isinstance(value, bool):
            raise TypeError(
                f"metric value must be numeric, got {type(value).__name__}"
            )
        entry = {"name": name, "value": value}
        if unit is not None:
            entry["unit"] = unit
        with self._lock:
            self._pending.append(entry)

    def flush(self):
        """Drain pending in batches; warn on failure, leave remainder for retry.

        Each batch is removed from `_pending` only after its
        put_metric_data succeeds. On exception, log a warning and
        return — a subsequent flush() (if the process lives long
        enough) gets another chance.
        """
        while True:
            with self._lock:
                if not self._pending:
                    return
                batch = self._pending[: self.BATCH_SIZE]

            try:
                Retryable.call(
                    lambda b=batch: self._client().put_metric_data(
                        Namespace=f"Turbofan/{self._pipeline_name}",
                        MetricData=[self._datum(e) for e in b],
                    ),
                    max_retry_seconds=None,
                )
            except Exception as exc:
                with self._lock:
                    remaining = len(self._pending)
                sys.stderr.write(
                    f"[turbofan_runtime] WARNING: failed to flush "
                    f"{remaining} metrics: {type(exc).__name__}: {exc}\n"
                )
                return

            with self._lock:
                self._pending = self._pending[len(batch):]

    def _client(self):
        if self._cw is not None:
            return self._cw
        with self._lock:
            if self._cw is None:
                # Disable SDK retries — Retryable owns the policy.
                self._cw = boto3.client(
                    "cloudwatch",
                    config=Config(
                        retries={"mode": "standard", "total_max_attempts": 1},
                        connect_timeout=10,
                        read_timeout=30,
                    ),
                )
            return self._cw

    def _datum(self, entry):
        datum = {
            "MetricName": entry["name"],
            "Value": entry["value"],
            "Dimensions": self._dimensions(),
        }
        if "unit" in entry:
            datum["Unit"] = entry["unit"]
        return datum

    def _dimensions(self):
        dims = [
            {"Name": "Pipeline", "Value": self._pipeline_name},
            {"Name": "Stage", "Value": self._stage},
            {"Name": "Step", "Value": self._step_name},
        ]
        if self._size:
            dims.append({"Name": "Size", "Value": str(self._size)})
        return dims
